src/Features/CSharp/Portable/ExtractMethod/CSharpSelectionValidator.cs

```diff
@@ -20,538 +18,574 @@
 
 namespace Microsoft.CodeAnalysis.CSharp.ExtractMethod;
 
-internal sealed partial class CSharpSelectionValidator(
-    SemanticDocument document,
-    TextSpan textSpan,
-    bool localFunction) : SelectionValidator<CSharpSelectionResult, StatementSyntax>(document, textSpan)
+internal sealed partial class CSharpExtractMethodService
 {
-    private readonly bool _localFunction = localFunction;
-
-    public override async Task<(CSharpSelectionResult, OperationStatus)> GetValidSelectionAsync(CancellationToken cancellationToken)
+    internal sealed partial class CSharpSelectionValidator(
+        SemanticDocument document,
+        TextSpan textSpan,
+        bool localFunction) : SelectionValidator(document, textSpan)
     {
-        if (!ContainsValidSelection)
-            return (null, OperationStatus.FailedWithUnknownReason);
-
-        var text = SemanticDocument.Text;
-        var root = SemanticDocument.Root;
-        var model = SemanticDocument.SemanticModel;
-        var doc = SemanticDocument;
-
-        // go through pipe line and calculate information about the user selection
-        var selectionInfo = GetInitialSelectionInfo(root, text);
-        selectionInfo = AssignInitialFinalTokens(selectionInfo, root, cancellationToken);
-        selectionInfo = AdjustFinalTokensBasedOnContext(selectionInfo, model, cancellationToken);
-        selectionInfo = AssignFinalSpan(selectionInfo, text);
-        selectionInfo = ApplySpecialCases(selectionInfo, text, SemanticDocument.SyntaxTree.Options, _localFunction);
-        selectionInfo = CheckErrorCasesAndAppendDescriptions(selectionInfo, root);
-
-        // there was a fatal error that we couldn't even do negative preview, return error result
-        if (selectionInfo.Status.Failed)
-            return (null, selectionInfo.Status);
-
-        var controlFlowSpan = GetControlFlowSpan(selectionInfo);
-        if (!selectionInfo.SelectionInExpression)
+        private readonly bool _localFunction = localFunction;
+
+        protected override bool AreStatementsInSameContainer(StatementSyntax statement1, StatementSyntax statement2)
```

Comment:
moved here from CSharpSyntaxFacts.  This is a feature-level helper, and should remain with teh feature.